chore(docs): Update opentelemetry example to show access logs

mmanciop opened the pull request:
Commit Message: Update the OpenTelemetry demo to show access log correlation and a more recent collector
Additional Description: This PR updates the OpenTelemetry demo to show how to send access logs to an OpenTelemetry collector, and how to correlate those logs with tracing data (pending #30268). Also, use a more recent build of the OpenTelemetry collector.
Risk Level: Low
Testing: Manual
Docs Changes: None
Release Notes: The OpenTelemetry demo has been updated to show access log correlation with tracing data and to use a more recent build of the OpenTelemetry collector
Platform Specific Features: N/A
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

phlax commented on examples/opentelemetry/Dockerfile-opentelemetry:
```diff
@@ -1,11 +1,9 @@
-USER 0
-RUN chmod o+r /etc/otel-collector-config.yaml
-USER nobody
+FROM otel/opentelemetry-collector:0.99.0@sha256:bba8474ea69d11f0450e38c50040efea47f4190723a9fb64cc9d44939b4ade4c
```
this is the same image

im not adverse to using the version explicitly tho

mmanciop replied:
Same image, different version. The collector is not stable yet. For example, the rename of `logging` to `debug` exporter that is commented on elsewhere in this PR.

phlax replied:
my point is the sha hasnt changed - ie bba8474ea69d11f0450e38c50040efea47f4190723a9fb64cc9d44939b4ade4c

when you set the sha docker pretty much ignores everything else anyway

mmanciop replied:
Yes, I see. It's a bit confusing IMO to use a `latest` with SHA-pinning, but frankly I can go either way.

phlax replied:
if you want to separate out that change i would be happy to review - i think its more explicit, and depending on how upstream releases can reduce update noise

phlax commented on examples/opentelemetry/Dockerfile-opentelemetry:
```diff
@@ -1,11 +1,9 @@
 FROM alpine:3.19@sha256:c5b1261d6d3e43071626931fc004f70149baeba2c8ec672bd4f27761f8e1ad6b as otelc_curl
-RUN apk --update add curl
+    RUN apk --update add curl
```
please unindent this

mmanciop replied:
Done in 6aedc5e4d1

phlax commented on examples/opentelemetry/Dockerfile-opentelemetry:
```diff
@@ -1,11 +1,9 @@
-RUN chmod o+r /etc/otel-collector-config.yaml
-USER nobody
+FROM otel/opentelemetry-collector:0.99.0@sha256:bba8474ea69d11f0450e38c50040efea47f4190723a9fb64cc9d44939b4ade4c
+    COPY --from=otelc_curl / /
+    COPY ./otel-collector-config.yaml /etc/otel-collector-config.yaml
+    USER 0
+    RUN chmod o+r /etc/otel-collector-config.yaml
+    USER nobody
```
please unindent these

mmanciop replied:
Done in 6aedc5e4d1

phlax commented on examples/opentelemetry/docker-compose.yaml:
```diff
@@ -63,9 +63,11 @@ services:
       test: ["CMD-SHELL", "curl -sf http://localhost:13133 || exit 1"]
       interval: 1s
-      timeout: 120s
+      timeout: 10s
```
im not sure why this change is here or whether its desirable - i think it would make it retry every 1/12s

that said - 120 retries is probably too high - not sure where these values came from originally

mmanciop replied:
The collector comes up in a second or two tops. 120s is IMO just cause for FUD for the end user.

phlax replied:
cool - probably in that case tho - my reading of docs (maybe incorrectly) is that we would want also to change the `retries` values also to prevent it spinning out with healthchecks - happy to be wrong about that

phlax replied:
ok - nm - i misread - `interval` is set

in which case i think `retries` is most likely not required at all, but im possibly missing something there

mmanciop replied:
> Interval
> 
> The health check will first run interval seconds after the container is started, and then again interval seconds after each previous check completes.
> 
> Timeout
> 
> If a single run of the check takes longer than timeout seconds then the check is considered to have failed.
> 
> Retries
> 
> It takes retries consecutive failures of the health check for the container to be considered unhealthy.

From [StackOverflow](https://www.google.com/url?sa=t&source=web&rct=j&opi=89978449&url=https://stackoverflow.com/questions/68186759/docker-compose-healthcheck-options&ved=2ahUKEwjXqrbV_-6FAxXqVfEDHThtD_0QFnoECB4QAQ&usg=AOvVaw3Jka0tjOljGCdCNLzBOugc) and [this cheatsheet](https://kapeli.com/cheat_sheets/Dockerfile.docset/Contents/Resources/Documents/index). I feel retries are useless. The collector will virtually always answer if it's alive (unless under the most massive GC I ever saw it doing), and if it's not alive, it's 99% a config issue, which are not helped by retries.

phlax replied:
thanks - i didnt read docs properly obviously - not sure in that case but what we want to avoid at all costs is making the sandbox flakey in ci

mmanciop replied:
I don't feel getting rid of retries in this case should make anything flakier tbh. If anything, it'd fail faster.

phlax commented on examples/opentelemetry/docker-compose.yaml:
```diff
@@ -63,9 +63,11 @@ services:
+    - "4137:4137" # gRPC port
     - "${PORT_UI:-55679}:55679"
+    volumes:
+    - ./otel-collector-config.yaml:/etc/otel-collector-config.yaml
```
this is unnecessary and breaks CI - the examples dont allow you to mount Envoy configs from the filesystem - the reason for this is that some devs work on a filesystem with strict `umask` set

mmanciop replied:
Ok

phlax replied:
just realized this isnt an envoy config but i think the same applies - it will need to be loaded in the docker build

mmanciop replied:
Done in 4c4e27eab5

phlax commented on examples/opentelemetry/envoy-1.yaml:
```diff
@@ -25,6 +25,16 @@ static_resources:
                     cluster_name: opentelemetry_collector
                   timeout: 0.250s
                 service_name: envoy-1
+          access_log:
```
i think these `access_log` changes are the real change in the PR

im not totally clear what the purpose is but if we are going to change these then the docs and `verify.sh` script should reflect that

mmanciop replied:
> i think these access_log changes are the real change in the PR

Pretty much, but also updating the collector to a more recent version and adjusting the configs is worthy in its own right.

phlax replied:
unless i misunderstand tho - the version isnt updated - just made explicit

happy to carry some cleanups, but i think the change is probably great and requires focus

to land, we really need to update the docs to explain, and relatedly to test what the docs say to prevent bitrot etc

mmanciop replied:
You are right, the version did not actually change. Before updating the docs, I would love for the correlation to... well actually work (see #30268). My idea to open the PR this early was to give @ashishb-solo a quick way to test the changes, and to motivate why the PR to actually fix the correlation was needed.

phlax replied:
cool, thanks for context

is it the same issue as https://github.com/envoyproxy/envoy/issues/33906 ?

i reopened the one you posted but i have a feeling that they are now duping

with these sandboxes they are documentation first, so to justify adding a load of configuration it needs to be documented and tested

mmanciop replied:
@phlax : #33906 is about the correlation not working. And it still doesn't, it's broken at a technical level. This demo will start working as expected, and showcase the correlation, when the underlying bug in the `opentelemetry` access log exporter is fixed :-)

mmanciop replied:
FYI #33909 solves the problem pretty nicely, see https://github.com/envoyproxy/envoy/pull/33909#issuecomment-2092412409

phlax commented on examples/opentelemetry/otel-collector-config.yaml:
```diff
@@ -4,6 +4,7 @@ extensions:
   zpages:
     endpoint: 0.0.0.0:55679
   health_check:
+    endpoint: 0.0.0.0:13133
```
why is this necessary?

mmanciop replied:
I originally did it because the default of the collector will change from binding to 0.0.0.0 to localhost, but then again, the health-check is run from within the container, so there is no actual difference.

mmanciop replied:
Reverted in 8e723eb5c8

phlax replied:
cool, i was curious about this more than anything

phlax commented on examples/opentelemetry/otel-collector-config.yaml:
```diff
@@ -21,18 +22,17 @@ processors:
       receivers: [otlp]
       processors: [memory_limiter, batch]
-      exporters: [logging]
+      exporters: [debug]
```
im not sure we want this - but also not sure iiuc the implications

mmanciop replied:
A renaming in the Otel collector that happens a few versions ago.

mmanciop replied:
<img width="806" alt="Screenshot 2024-05-02 at 14 31 07" src="https://github.com/envoyproxy/envoy/assets/14020598/e74b546e-987d-466c-b728-e712a9a33092">

https://github.com/open-telemetry/opentelemetry-collector/tree/main/exporter/loggingexporter

phlax replied:
ok, thanks for explanation

phlax reviewed (commented):
thanks for working on this @mmanciop

can i suggest removing all changes other than the access log additions

as said below, if we make the access log changes then this should be reflected in the docs for the sandbox, and tested

phlax commented:
/docs

phlax reviewed (commented):
this is the file that needs to be edited to update docs https://github.com/envoyproxy/envoy/blob/main/docs/root/start/sandboxes/opentelemetry.rst

i think lets insert a new step 4 (maybe more) that demonstrates what the config is doing

and this is the rendered docs for this sandbox at current commit https://storage.googleapis.com/envoy-pr/a712ed7/docs/start/sandboxes/opentelemetry.html

once we update we can look at the https://github.com/envoyproxy/envoy/blob/main/docs/root/start/sandboxes/verify.sh which is intended to ~reflect the steps in the docs

mmanciop commented:
> this is the file that needs to be edited to update docs https://github.com/envoyproxy/envoy/blob/main/docs/root/start/sandboxes/opentelemetry.rst
> 
> i think lets insert a new step 4 (maybe more) that demonstrates what the config is doing
> 
> and this is the rendered docs for this sandbox at current commit https://storage.googleapis.com/envoy-pr/a712ed7/docs/start/sandboxes/opentelemetry.html
> 
> once we update we can look at the https://github.com/envoyproxy/envoy/blob/main/examples/opentelemetry/verify.sh which is intended to ~reflect the steps in the docs

Done in 8abd1e33331da13c4a91f8a79ed2f9a488bdb99f

mmanciop commented:
Preparing the docs I noticed that maybe a little more setup is needed to add attributes. It will take me a little to hammer out :-)

mmanciop commented:
/docs

phlax commented:
@mmanciop will review more thoroughly shortly - but copying out ci suggestion to get past (one of) ci fails

```console
  Please fix your editor to ensure:

      - no trailing whitespace
      - no preceding mixed tabs/spaces
      - all files end with a newline
```

phlax commented on docs/root/start/sandboxes/opentelemetry.rst:
```diff
@@ -119,3 +120,157 @@ In the ``Latency Samples`` of ``opentelemetry.proto.collector.trace.v1.TraceServ
+
+Set up a cluster to report to the OpenTelemetry collector (or another endpoint that accepts OTLP v1):
+
+.. code-block:: yaml
```
this is failing when trying to build the rst

i _think_ the reason is that its confused by the yaml beneath - thinking its an rst directive/argument

i would suggest leaving a space after all `.. code-block:: ...` lines

in this particular case its an Envoy config - so it needs to go in a full bootstrap config and included - you can find similar examples with `git grep literalinclude docs/`

phlax replied:
(the rendered CI `/docs` will show for the last successful build and not at this commit)

phlax replied:
worth mentioning - you should be able to `literalinclude` the actual example configs 

in which case `git grep literalinclude docs/root/start/sandboxes` is probably more helpful

ill wait until the docs are rendering before reviewing again

mmanciop replied:
Thanks for the pointers, I gave it a much better try with 051d66012b

phlax commented on docs/root/start/sandboxes/_include/envoy-opentelemetry.yaml:
```diff
@@ -0,0 +1,120 @@
+static_resources:
```
we probably dont need to add this - we can use the file here (eg) https://github.com/envoyproxy/envoy/blob/main/examples/opentelemetry/envoy-front-proxy.yaml with `_include/opentelemetry/envoy-front-proxy.yaml`

mmanciop replied:
Done in d01a66c186

phlax commented on docs/root/start/sandboxes/opentelemetry.rst:
```diff
@@ -119,3 +120,132 @@ In the ``Latency Samples`` of ``opentelemetry.proto.collector.trace.v1.TraceServ
+    opentelemetry-opentelemetry-1      | Flags: 0
+    opentelemetry-opentelemetry-1      |    {"kind": "exporter", "data_type": "logs", "name": "debug"}
+
+Notice how the log record's `Trace ID` and `Span ID` fields match the span's `Trace ID` and `ID` fields.
```
```suggestion
Notice how the log record's ``Trace ID`` and ``Span ID`` fields match the span's ``Trace ID`` and ``ID`` fields.
```

its kinda annoying in some respects but `literals` in rst are ` `` `

phlax commented on examples/opentelemetry/docker-compose.yaml:
```diff
@@ -63,9 +63,9 @@ services:
-    command: ["--config=/etc/otel-collector-config.yaml"]
+      start_period: 2s
     ports:
+    - "4137:4137" # gRPC port
```
i suspect this isnt required it exposes the port to the host and would usually be referenced in the docs in that case

mmanciop replied:
Fair

mmanciop replied:
Done in fc93b3573d

phlax commented on examples/opentelemetry/envoy-1.yaml:
```diff
@@ -4,6 +4,48 @@
 #   10001 -> routes to `envoy-2`
 
 static_resources:
+  clusters:
```
please keep clusters _after_ listeners as that is consistent with the existing codebase

mmanciop replied:
But then I cannot use YAML anchors in a non-confusing way :-(

mmanciop replied:
Done in 348c1dc0e1

mmanciop replied:
Done in 91f0514864232a285b9ec9ef220d00974d4f3705

phlax reviewed (commented):
thanks for iterating @mmanciop - adding/updating these sandboxes generally takes a few

main feedback is to use the existing yaml files from the `examples/opentelemtry` dir as any user checking out repo and following along will have them directly

these get added automagically to the rst build dir (by bazel) so can be referenced with `_include/openetelemetry/*yaml`

existing sandbox docs mostly dont have captions but elsewhere we use this for downloading the file - in that case i think its "name" should be `examples/opentelemetry/*.yaml` - ie repo-relative

phlax reviewed (commented):
@mmanciop quick docs review

step 5 comes after the `see also` section which should be last - i would move step 5 -> step 4 and increment current step 5

wrt step 6 - its misplaced here - this should be in the config reference section - but when i checked that it seems that this page is currently missing or ~empty - this is the only config reference i could see https://www.envoyproxy.io/docs/envoy/latest/configuration/observability/stat_sinks/open_telemetry_stat_sink.html

mmanciop commented:
> @mmanciop quick docs review
> 
> step 5 comes after the `see also` section which should be last - i would move step 5 -> step 4 and increment current step 5
> 
> wrt step 6 - its misplaced here - this should be in the config reference section - but when i checked that it seems that this page is currently missing or ~empty - this is the only config reference i could see https://www.envoyproxy.io/docs/envoy/latest/configuration/observability/stat_sinks/open_telemetry_stat_sink.html

Moved see_also to the end of the page in 62e5b84f5f, and deleted the configuration section that is misplaced.

phlax commented on docs/root/start/sandboxes/opentelemetry.rst:
```diff
@@ -112,6 +113,81 @@ In the ``Latency Samples`` of ``opentelemetry.proto.collector.trace.v1.TraceServ
+Check the spans Docker-compose logs:
+
+.. code-block:: console
+   :linenos:
```
lets remove the linenos from the console dump - its meant more for files so users can see that its coming from a file and match the lines

mmanciop replied:
Done in b4a96fe0f5875d108a49b54fa30720273c35dcb8

phlax commented on docs/root/start/sandboxes/opentelemetry.rst:
```diff
@@ -112,6 +113,81 @@ In the ``Latency Samples`` of ``opentelemetry.proto.collector.trace.v1.TraceServ
 
 .. image:: /start/sandboxes/_static/opentelemetry-ui-traces.png
 
+Step 5: View the access logs in the OpenTelemetry collector output
```
could you swap step 4 and step 5 please - i think the screenshots should be the last step

mmanciop replied:
Done in b4a96fe0f5875d108a49b54fa30720273c35dcb8

phlax commented on docs/root/start/sandboxes/opentelemetry.rst:
```diff
@@ -112,6 +113,81 @@ In the ``Latency Samples`` of ``opentelemetry.proto.collector.trace.v1.TraceServ
+Step 5: View the access logs in the OpenTelemetry collector output
+******************************************************************
+
+Check the spans Docker-compose logs:
```
```suggestion
Check the spans Docker Compose logs:
```

i think we want to be explicit about how to do this - i would add as the first line in the console logs below something like:

```console
$ docker compose logs opentelemetry
```

mmanciop replied:
Done in b4a96fe0f5875d108a49b54fa30720273c35dcb8

phlax commented on docs/root/start/sandboxes/opentelemetry.rst:
```diff
@@ -112,6 +113,81 @@ In the ``Latency Samples`` of ``opentelemetry.proto.collector.trace.v1.TraceServ
+    opentelemetry-opentelemetry-1      |      -> response_flags: Str(-)
+    opentelemetry-opentelemetry-1      |    {"kind": "exporter", "data_type": "traces", "name": "debug"}
+
+And the matching access log:
```
in this case im a bit unclear if this is the same (container) log as above

mmanciop replied:
They are from the same container. Unfortunately the access logs do not have resource detection working and don't set the `service.name` resource attribute. It's one of the other shortcomings that the Otel support for access logs has.

phlax reviewed (commented):
really close, thanks @mmanciop 

for the `verify.sh` script i think we just want to add another section reflecting the addition to the docs - the tests can be pretty simple - just grep in the logs for the highlighted lines - something like

```console
"${DOCKER_COMPOSE[@]}" logs opentelemetry | grep "Trace ID"
```

phlax commented:
@mmanciop could you update the verify.sh script and fix formatting so we can see the verify ci running

for ref you can run/test locally (assumes local user has access to Docker socket)

```console
$ cd examples/opentelemetry
$ ./verify.sh
```

mmanciop commented:
> @mmanciop could you update the verify.sh script and fix formatting so we can see the verify ci running
> 
> 
> 
> for ref you can run/test locally (assumes local user has access to Docker socket)
> 
> 
> 
> ```console
> 
> $ cd examples/opentelemetry
> 
> $ ./verify.sh
> 
> ```

@phlax that's what I am working on, but it won't be pretty. The collector debug output cannot currently be formatted to JSON and that's gonna hurt.

phlax commented:
in that case just use grep i reckon - that is ~equivalent to the highlighted lines above i think

phlax commented:
for a stricter test you could fish out the ids and check they match - that would be pretty cool - but i reckon optional

mmanciop commented:
> in that case just use grep i reckon - that is ~equivalent to the highlighted lines in console output docs i think

Multiline `grep` is a portability nightmare in my experience. I cobbled up something that I think is reasonable in 0a71a08107. It won't succeed off until #33909 is merged and this PR is rebased on top of it.

phlax commented:
/wait for https://github.com/envoyproxy/envoy/pull/33909
